sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +131,77 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    // As the analyzer makes sure that the list of In is already of the same data type, then the
+    // rule can simply check the first literal in `in.list` can implicitly cast to `toType` or not,
+    // and note that:
+    // 1. this rule doesn't convert in when `in.list` is empty.
+    // 2. this rule only handles the case when both `fromExp` and value in `in.list` are of numeric
+    // type.
+    case in @ In(Cast(fromExp, toType: NumericType, _), list @ Seq(firstLit, _*))
+        if canImplicitlyCast(fromExp, toType, firstLit.dataType) && in.inSetConvertible =>
+      val (newValueList, expr) =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .partition {
+            case EqualTo(_, _: Literal) => true
+            case And(IsNull(_), Literal(null, BooleanType)) => false
+            case _ => throw new IllegalStateException("Illegal unwrap cast result found.")
+          }
+
+      val (nonNullValueList, nullValueList) = newValueList.partition {
+        case EqualTo(_, NonNullLiteral(_, _: NumericType)) => true
+        case EqualTo(_, Literal(null, _)) => false
+        case _ => throw new IllegalStateException("Illegal unwrap cast result found.")
+      }
+      // make sure the new return list have the same dataType.
+      val newList = {
+        if (nonNullValueList.nonEmpty) {
+          // cast the null value to the dataType of nonNullValueList
+          // when the nonNullValueList is nonEmpty.
+          nullValueList.map {
+            case EqualTo(_, lit) =>
+              Cast(lit, nonNullValueList.head.asInstanceOf[EqualTo].left.dataType)
+          } ++ nonNullValueList.map {case EqualTo(_, lit) => lit}
+        } else {
+          // the new value list only contains null value,
+          // cast the null value to fromExp.dataType.
+          nullValueList.map {
+            case EqualTo(_, lit) =>
+              Cast(lit, fromExp.dataType)
```

Comment:
Why do we need a Cast? The returned literal should already be `fromExp.dataType`

Reply:
The same with `nonNullValueList.nonEmpty`, when the in.list only contains value `null`, the `unwrapCast` returns the original expression and the original literal.datatype (same with toType) may difference from `fromExp.dataType`. We should make sure the value in in.list has the same data type as in.value.